SPEC.md

```diff
@@ -153,6 +154,47 @@ The network configuration is described in JSON form. The configuration can be st
   - `options` (list of strings): list of options that can be passed to the resolver
 
 Plugins may define additional fields that they accept and may generate an error if called with unknown fields. The exception to this is the `args` field may be used to pass arbitrary data which may be ignored by plugins.
+
+### Plugin Chaining and Delegation
+
+Multiple plugins may be chained together to perform more complex network configuration. IPAM plugins already operate this way, with the outer plugin delegating to the IPAM plugin and receiving its IPAM result.
+The network configuration may contain an `inner` field, which itself is a complete CNI network configuration block (except for the `cniVersion` and `name` fields, which cannot be specified in the `inner` dictionary).
+When the outer plugin finds the `inner` field, it proceeds to execute the inner plugin, passing the entire `inner` configuration via stdin, merging the top-level `cniVersion` and `name` fields.
+The inner plugin then executes, including possibly running yet another inner plugin if the top-level `inner` dictionary contains a second `inner` dictionary, specifying a third plugin.
+Each plugin then receives the IPAM result of the called plugin, and can use that result and optionally manipulate it before returning the result to its caller.
+
+For example, suppose traffic shaping should be enabled for a container whose network is configured with the standard `bridge` plugin and `host-local` IPAM plugins.
+
+```json
+{
+  "cniVersion": "0.2.0",
+  "name": "chaining-example",
+  "type": "traffic-shaping",
+  "ingressbw": "20M",
+  "egressbw": "10M",
```

Comment:
For k8s, @thockin wanted the CNI network configs to be as kubernetes agnostic as possible.  That meant that we have a mostly static CNI network config like what I list here, so that plugins don't have to know kubernetes is running the show.

What I proposed instead in the kube hostports issue was CNI "template" files that kubernetes would read and interpret, then replace certain fields at runtime so that when the CNI plugin was called, it had a complete static config on stdin.

The template approach would allow plugin writers to create kube-agnostic network plugin chains (well, except for pod name/namespace in CNI_ARGS), making on the template files container-runtime-specific.